### DISCUSSION TITLE
[Feature] Allow Multiple URLs for the Machine Learning Server

### DISCUSSION THREAD

**Discussion by akropp** [immich-app/immich#4616] [Feature Request]
### The feature

To help speed the process of running a large number of ML tasks (facial recognition, tagging, etc), particularly when importing a new library or re-running the facial recognition job, it would be useful to be able to run multiple instances of the machine learning server on different machines, and then have immich_server load-balance across them.  

Simplest would be to just provide a list of hosts, and the server could round-robin among them, allowing for the possibility that a host in the list may be down/unavailable.  An improvement would be to either have some sort of weighting per host, or a max concurrent jobs per host.

### Platform

- [X] Server
- [X] Web
- [ ] Mobile

**Comment by ddshd** [comment]
I’ve currently done this though putting nginx load balancing in front of my machine learning hosts. I point Immich to the LB url and the LB will send the traffic to different hosts

**Reply by macross5** [reply]
It worked for me, but only if i set a **real ip-address**. Something like http://xyz.there working for the jobs but not for the smart search in the web-client. Tried it with caddy, where adguard sends the traffic from *.there to caddy.

**Reply by BloodBlight** [reply]
It also worked for me using the IP address of my docker swarm VIP.

**Comment by mertalev** [comment]
I think this comes down to user setup; it feels a bit out of scope for Immich.

**Comment by BloodBlight** [comment]
Not to revive a dead topic, but it might be relevant for future readers (as I am also here for this reason).

Based on this point in the code:
https://github.com/immich-app/immich/blob/3901b5da44b347242ceb964e9c2dc13c9d5912d6/docs/docs/administration/system-settings.md?plain=1#L22C1-L25C4

Referencing this video for clustering and DBSCAN:
https://www.youtube.com/watch?v=RDZUdRSDOok

Would placing the AI service behind a load balancer not impact the results quality?  If I understand it correctly, it would still work, but you might have greater segmentation of results (you will have to merge more sets into a single person)?  Just speculation from my 10 minute watching of a YouTube video though...

And expert input would be welcomed!

**Reply by mertalev** [reply]
Actually no, because the machine learning service is only responsible for generating the embeddings. It doesn't handle the clustering. But scaling up the number of microservices instances would somewhat impact the clustering because the concurrency is only fixed at 1 at an instance level, not global.

**Reply by BloodBlight** [reply]
Awesome, thank you for the clarification, I will have to try this on my docker swarm!  :)
